File: benchmarks/select_fanin.go
package main

import (
	"fmt"
	"sync"
	"time"
)

const perSource = 2048

func produce(tx chan<- int, start int, done *sync.WaitGroup) {
	defer done.Done()
	for i := 0; i < perSource; i++ {
		tx <- start + i*2
	}
}

// receivePriority matches chan.select's deterministic tie rule: if both
// receivers are ready, the first wins. The second select blocks only when the
// first receiver is empty.
func receivePriority(a, b <-chan int) (int, int) {
	select {
	case value := <-a:
		return 0, value
	default:
	}
	select {
	case value := <-a:
		return 0, value
	case value := <-b:
		return 1, value
	}
}

func expectedChecksum() int {
	checksum := 0
	for i := 0; i < perSource; i++ {
		checksum = (checksum*257 + i*2 + 1) % 1000000007
	}
	for i := 0; i < perSource; i++ {
		checksum = (checksum*257 + i*2 + 3) % 1000000007
	}
	return checksum
}

func main() {
	a := make(chan int, perSource)
	b := make(chan int, perSource)
	var producers sync.WaitGroup
	producers.Add(2)
	go produce(a, 0, &producers)
	go produce(b, 1, &producers)
	producers.Wait()

	seenA, seenB := 0, 0
	checksum := 0
	orderOK := true
	t0 := time.Now()
	for seenA+seenB < perSource*2 {
		source, value := receivePriority(a, b)
		if source == 0 {
			if seenB != 0 || value != seenA*2 {
				orderOK = false
			}
			checksum = (checksum*257 + value + 1) % 1000000007
			seenA++
		} else {
			if seenA != perSource || value != seenB*2+1 {
				orderOK = false
			}
			checksum = (checksum*257 + value + 2) % 1000000007
			seenB++
		}
	}
	benchNS := time.Since(t0).Nanoseconds()

	expected := expectedChecksum()
	if orderOK && seenA == perSource && seenB == perSource && checksum == expected {
		fmt.Printf("count=%d checksum=%d order=ok\n", seenA+seenB, checksum)
	} else {
		fmt.Printf("count=%d checksum=%d expected=%d order=bad\n", seenA+seenB, checksum, expected)
	}
	fmt.Printf("bench_ns=%d\n", benchNS)
}

File: benchmarks/select_fanin.rs
use std::sync::mpsc;
use std::thread;
use std::time::Instant;

const PER_SOURCE: i32 = 2048;

fn expected_checksum() -> i32 {
    let mut checksum = 0i64;
    for i in 0..PER_SOURCE {
        checksum = (checksum * 257 + (i * 2 + 1) as i64) % 1000000007;
    }
    for i in 0..PER_SOURCE {
        checksum = (checksum * 257 + (i * 2 + 3) as i64) % 1000000007;
    }
    checksum as i32
}

fn main() {
    let (tx_a, rx_a) = mpsc::sync_channel(PER_SOURCE as usize);
    let (tx_b, rx_b) = mpsc::sync_channel(PER_SOURCE as usize);
    
    let handle_a = thread::spawn(move || {
        for i in 0..PER_SOURCE {
            tx_a.send(i * 2).unwrap();
        }
    });
    
    let handle_b = thread::spawn(move || {
        for i in 0..PER_SOURCE {
            tx_b.send(1 + i * 2).unwrap();
        }
    });
    
    handle_a.join().unwrap();
    handle_b.join().unwrap();

    let mut seen_a = 0;
    let mut seen_b = 0;
    let mut checksum = 0i64;
    let mut order_ok = true;
    
    let t0 = Instant::now();
    
    // In Rust we can just try to receive from a, then if empty try b, etc.
    // However mpsc::select! is deprecated/removed in std, so we use try_recv
    while seen_a + seen_b < PER_SOURCE * 2 {
        if let Ok(value) = rx_a.try_recv() {
            if seen_b != 0 || value != seen_a * 2 {
                order_ok = false;
            }
            checksum = (checksum * 257 + value as i64 + 1) % 1000000007;
            seen_a += 1;
        } else if let Ok(value) = rx_b.try_recv() {
            if seen_a != PER_SOURCE || value != seen_b * 2 + 1 {
                order_ok = false;
            }
            checksum = (checksum * 257 + value as i64 + 2) % 1000000007;
            seen_b += 1;
        }
    }
    
    let bench_ns = t0.elapsed().as_nanos();
    let expected = expected_checksum();
    
    if order_ok && seen_a == PER_SOURCE && seen_b == PER_SOURCE && checksum == expected as i64 {
        println!("count={} checksum={} order=ok", seen_a + seen_b, checksum);
    } else {
        println!("count={} checksum={} expected={} order=bad", seen_a + seen_b, checksum, expected);
    }
    println!("bench_ns={}", bench_ns);
}
